Client/Frontend/Widgets/AutocompleteTextField.swift

```diff
@@ -82,27 +87,51 @@ class AutocompleteTextField: UITextField, UITextFieldDelegate {
         }
     }
 
+    // `shouldChangeCharactersInRange` is called before the text changes, and SELtextDidChange is called after.
+    // Since the text has changed, remove the completion here, and SELtextDidChange will fire the callback to
+    // get the new autocompletion.
     func textField(textField: UITextField, shouldChangeCharactersInRange range: NSRange, replacementString string: String) -> Bool {
-        if completionActive && string.isEmpty {
-            // Characters are being deleted, so clear the autocompletion, but don't change the text.
-            removeCompletion()
-            return false
+        if completionActive {
+            if string.isEmpty {
+                // Characters are being deleted, so clear the autocompletion, but don't change the text.
+                removeCompletion()
+                return false
+            }
+            removeCompletionIfRequiredForEnteredString(string)
         }
-
         return true
     }
 
+    private func  removeCompletionIfRequiredForEnteredString(string: String) {
+        let actualEnteredString = text.substringToIndex(advance(text.startIndex, enteredTextLength, text.endIndex)) + string
+        // If actual entred text by user not starts with previous suggestion than remove completion.
+        if !previousSuggestion.startsWith(actualEnteredString) {
+            removeCompletion()
+        }
+        enteredTextLength += count(string)
+    }
+
     func setAutocompleteSuggestion(suggestion: String?) {
-        if let suggestion = suggestion where editing {
+        // Setting auto complete suggestion in between breaks input sesssion. For multiple input characters 
+        // while text is marked than user is not entered its full input. So here we are checking `markedTextRange`. 
+        // If it is nil it means now multiinput character session is completed and these character can not change.
```

Comment:
Some adjustments:
```
// Setting the autocomplete suggestion during multi-stage input will break the session since the text
// is not fully entered. If `markedTextRange` is nil, that means the multi-stage input is complete, so
// it's safe to append the suggestion.
```